heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/MetricsCacheQueryUtils.java

```diff
@@ -0,0 +1,290 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.twitter.heron.metricscachemgr.metricscache;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import com.twitter.heron.metricscachemgr.metricscache.datapoint.ExceptionDatapoint;
+import com.twitter.heron.proto.system.Common;
+import com.twitter.heron.proto.tmaster.TopologyMaster;
+
+/**
+ * define the query request and response format
+ */
+public final class MetricsCacheQueryUtils {
+  // logger
+  private static final Logger LOG = Logger.getLogger(MetricsCacheQueryUtils.class.getName());
+
+  private MetricsCacheQueryUtils() {
+  }
+
+  /**
+   * compatible with com.twitter.heron.proto.tmaster.TopologyMaster.MetricRequest
+   */
+  public static MetricRequest Convert(TopologyMaster.MetricRequest request) {
+    String componentName = request.getComponentName();
+
+    MetricRequest request1 = new MetricRequest();
+    request1.componentNameInstanceId = new HashMap<>();
+    if (request.getInstanceIdCount() == 0) {
+      // empty list means all instances
+      // 'null' means all instances
+      request1.componentNameInstanceId.put(componentName, null);
+    } else {
+      Set<String> instances = new HashSet<>();
+      // only one component
+      request1.componentNameInstanceId.put(componentName, instances);
+      // if there are instances specified
+      instances.addAll(request.getInstanceIdList());
+    }
+
+    Set<String> metrics = new HashSet<>();
+    request1.metricNames = metrics;
+    if (request.getMetricCount() > 0) {
+      metrics.addAll(request.getMetricList());
+    } // empty list means no metrics
+
+    // default: the whole time horizon
+    request1.startTime = 0;
+    request1.endTime = Long.MAX_VALUE;
+    if (request.hasInterval()) { // endTime = now
+      request1.endTime = System.currentTimeMillis();
+
+      long interval = request.getInterval(); // in seconds
+      if (interval <= 0) { // means all
+        request1.startTime = 0;
+      } else { // means [-interval, now]
+        request1.startTime = request1.endTime - interval * 1000;
+      }
+    } else {
+      request1.startTime = request.getExplicitInterval().getStart() * 1000;
+      request1.endTime = request.getExplicitInterval().getEnd() * 1000;
+    }
+
+    // default: aggregate all metrics
+    request1.minutely = 0;
```

Comment:
minutely is better named to 'aggregation granularity'
0: default, aggregate all metrics
1: aggregate metrics by bucket
2: no aggregation; return raw metrics